File: worker/governance_verification/binding.py
"""Internal task authority requires an immutable human request and exact context."""

import re

from worker.orchestration.db import WorkbenchError, content_hash, instant
from .checker import CHECK_ID, strict_json
from .source import verify_context


COMMAND = "governance_verification_v2"
ACTOR = "system:governance-verifier-v2"


def request_binding(connection, command, *, verify_source=True):
    try:
        request = connection.execute("SELECT * FROM verification_requests WHERE id=?", (command["id"],)).fetchone()
        if request is None:
            raise ValueError("request")
        request = dict(request)
        payload = strict_json(command["payload_json"])
        expected = {"schema_version": "verification-request-v2", "verification_request_id": request["id"],
                    "portfolio_id": request["portfolio_id"], "check_id": CHECK_ID, "context_hash": request["context_hash"]}
        if (command["command_type"] != COMMAND or command["actor_id"] != ACTOR or payload != expected
                or content_hash(payload) != command["payload_hash"] or command["portfolio_id"] != request["portfolio_id"]
                or command["created_at"] != request["requested_at"] or request["check_id"] != CHECK_ID
                or not isinstance(request["requested_by"], str) or not re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9_.:-]{0,159}", request["requested_by"])
                or re.match(r"system(?::|$)", request["requested_by"], re.IGNORECASE)):
            raise ValueError("command")
        context = strict_json(request["context_json"])
        verify_context(context, request["portfolio_id"], request["check_id"], request["context_hash"], verify_source=verify_source)
        audit = connection.execute("SELECT * FROM audit_events WHERE id=?", (request["audit_id"],)).fetchone()
        if (audit is None or audit["action"] != "request_verification" or audit["object_type"] != "verification_request"
                or audit["object_id"] != request["id"] or audit["portfolio_id"] != request["portfolio_id"]
                or audit["actor_id"] != request["requested_by"] or audit["created_at"] != request["requested_at"]
                or audit["ledger_revision"] is not None):
            raise ValueError("audit")
        proof = strict_json(audit["payload_json"])
        if (not isinstance(proof, dict) or set(proof) != {"actor_kind", "input", "result"} or proof["actor_kind"] != "human"
                or set(proof["input"]) != {"portfolio_id", "check_id", "expected_context_hash", "reason", "idempotency_key"}
                or proof["input"]["portfolio_id"] != request["portfolio_id"] or proof["input"]["check_id"] != CHECK_ID
                or proof["input"]["expected_context_hash"] != request["context_hash"]
                or proof["input"]["idempotency_key"] != command["idempotency_key"]
                or not isinstance(proof["input"]["reason"], str) or not 1 <= len(proof["input"]["reason"].strip()) <= 1000
                or proof["input"]["reason"] != proof["input"]["reason"].strip() or re.search(r"[\x00\x1c-\x1f\x85]", proof["input"]["reason"])
                or not isinstance(proof["input"]["idempotency_key"], str)
                or not re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9_.:-]{0,159}", proof["input"]["idempotency_key"])
                or proof["result"] != {"request_id": request["id"], "check_id": CHECK_ID,
                                       "context_hash": request["context_hash"], "status": "queued"}):
            raise ValueError("human")
        return request, context
    except (ValueError, TypeError, KeyError, IndexError) as exc:
        raise WorkbenchError("VERIFICATION_REQUEST_INVALID") from exc


def job_binding(connection, job, lease, *, verify_source=True):
    command = connection.execute("SELECT * FROM command_requests WHERE id=?", (job["command_request_id"],)).fetchone()
    if command is None:
        raise WorkbenchError("VERIFICATION_JOB_BINDING_INVALID")
    request, context = request_binding(connection, command, verify_source=verify_source)
    attempt = connection.execute("SELECT * FROM job_attempts WHERE job_id=? AND attempt=? AND fencing_token=?",
                                 (job["id"], lease.attempt, lease.fencing_token)).fetchone()
    if (job["id"] != lease.job_id or job["job_type"] != COMMAND or job["scope"] != request["portfolio_id"]
            or job["period"] != request["requested_at"][:10] or job["input_version"] != request["id"] + ":" + command["payload_hash"]
            or job["max_attempts"] != 3 or job["attempt_count"] != lease.attempt or job["fencing_token"] != lease.fencing_token
            or job["status"] != "running" or job["lease_owner"] != lease.owner
            or attempt is None or attempt["status"] != "running" or attempt["finished_at"] is not None
            or attempt["error_json"] is not None or instant(attempt["started_at"]) < instant(request["requested_at"])):
        raise WorkbenchError("VERIFICATION_JOB_BINDING_INVALID")
    return request, context, dict(attempt)

File: web/src/server/verifications/binding.ts
import type Database from "better-sqlite3";
import { canonical, hash } from "../ledger/service";
import { parseStrictJson } from "../strict-json";
import { verificationCommandSchema } from "./schemas";
import { parseVerificationContext } from "./source";
import { VERIFICATION_CHECK_ID, type VerificationCommand, type VerificationContext } from "./types";

export interface VerificationRequestRow {
  id: string; portfolio_id: string; check_id: string; context_json: string; context_hash: string;
  requested_by: string; audit_id: string; requested_at: string;
}
export const evidenceInvalid = (): Error => new Error("VERIFICATION_EVIDENCE_INVALID");
export function verificationInstant(value: string): string {
  const match = /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})(?:\.(\d{1,6}))?Z$/u.exec(value);
  if (!match) throw evidenceInvalid();
  const date = new Date(`${match[1]}.000Z`);
  if (!Number.isFinite(date.getTime()) || date.toISOString().slice(0, 19) !== match[1]) throw evidenceInvalid();
  return `${match[1]}.${(match[2] ?? "").padEnd(6, "0")}Z`;
}
export function readVerificationRequest(db: Database.Database, portfolio: string, requestId: string): { row: VerificationRequestRow; context: VerificationContext; input: VerificationCommand } {
  const row = db.prepare("SELECT * FROM verification_requests WHERE id=? AND portfolio_id=?").get(requestId, portfolio) as VerificationRequestRow | undefined;
  if (!row) throw new Error("VERIFICATION_REQUEST_NOT_FOUND");
  try {
    const context = parseVerificationContext(row.context_json);
    if (row.context_json !== canonical(context) || hash(context) !== row.context_hash || row.check_id !== VERIFICATION_CHECK_ID
      || context.portfolio_id !== portfolio || context.check_id !== row.check_id || row.requested_at !== verificationInstant(row.requested_at)
      || !/^[A-Za-z0-9][A-Za-z0-9_.:-]{0,159}$/u.test(row.requested_by) || /^system(?::|$)/iu.test(row.requested_by)) throw evidenceInvalid();
    const command = db.prepare("SELECT * FROM command_requests WHERE id=?").get(row.id) as { portfolio_id: string; command_type: string; payload_json: string; payload_hash: string; actor_id: string; created_at: string; idempotency_key: string } | undefined;
    const payload = { schema_version: "verification-request-v2", verification_request_id: row.id, portfolio_id: portfolio, check_id: row.check_id, context_hash: row.context_hash };
    if (!command || command.command_type !== "governance_verification_v2" || command.portfolio_id !== portfolio || command.actor_id !== "system:governance-verifier-v2"
      || command.created_at !== row.requested_at || command.payload_json !== canonical(payload) || command.payload_hash !== hash(payload)) throw evidenceInvalid();
    const audit = db.prepare("SELECT * FROM audit_events WHERE id=?").get(row.audit_id) as { actor_id: string; action: string; object_type: string; object_id: string; portfolio_id: string; created_at: string; payload_json: string; ledger_revision: number | null } | undefined;
    if (!audit || audit.actor_id !== row.requested_by || audit.action !== "request_verification" || audit.object_type !== "verification_request"
      || audit.object_id !== row.id || audit.portfolio_id !== portfolio || audit.created_at !== row.requested_at || audit.ledger_revision !== null) throw evidenceInvalid();
    const proof = parseStrictJson(audit.payload_json) as { actor_kind: string; input: unknown; result: unknown };
    const input = verificationCommandSchema.parse(proof.input);
    const result = { request_id: row.id, check_id: row.check_id, context_hash: row.context_hash, status: "queued" };
    if (canonical(proof) !== canonical({ actor_kind: "human", input, result }) || proof.actor_kind !== "human"
      || input.portfolio_id !== portfolio || input.check_id !== row.check_id || input.expected_context_hash !== row.context_hash
      || input.idempotency_key !== command.idempotency_key) throw evidenceInvalid();
    return { row, context, input };
  } catch { throw evidenceInvalid(); }
}
